coreneuron/logging/logging.hpp

```diff
@@ -0,0 +1,477 @@
+#ifndef _CORENEURONT__LOGGING_HPP_
+#define _CORENEURONT__LOGGING_HPP_
+
+#include <iomanip>
+#ifndef NDEBUG
+  #include <iostream>  // used by CORENEURON::log::test_logging_levels()
+#endif
+#include <string>
+using namespace std;
+
+#ifdef WITH_MPI
+  #include <sstream>
+  #include <mpi.h>
+#endif
+
+#include "CORENEURON/config.hpp"
+
+
+/**
+ * General variables for coloured terminal output.
+ *
+ */
+struct OUT
+{
+  public:
+    //! @{
+    static const string black;
+    static const string red;
+    static const string green;
+    static const string yellow;
+    static const string blue;
+    static const string magenta;
+    static const string cyan;
+    static const string white;
+    //! @}
+
+    //! @{
+    static const string bold;
+    static const string underline;
+    //! @}
+
+    //! resets both, colour and style
+    static const string reset;
+};
+
+const string OUT::black = "\033[30m";
+const string OUT::red = "\033[31m";
+const string OUT::green = "\033[32m";
+const string OUT::yellow = "\033[33m";
+const string OUT::blue = "\033[34m";
+const string OUT::magenta = "\033[35m";
+const string OUT::cyan = "\033[36m";
+const string OUT::white = "\033[37m";
+
+const string OUT::bold = "\033[1m";
+const string OUT::underline = "\033[4m";
+
+const string OUT::reset = "\033[0m";
+
+// enable easy logging of STL containers
+#define ELPP_STL_LOGGING
+// disable creation of default log file
+#define ELPP_NO_DEFAULT_LOG_FILE
+// enable passing `--logging-flags` via command line
+#define ELPP_LOGGING_FLAGS_FROM_ARG
+
+#ifndef NDEBUG
+  #define ELPP_DEBUG_ASSERT_FAILURE
+  #define ELPP_STACKTRACE_ON_CRASH
+#endif
+
+#ifdef CORENEURONT_NO_LOGGING
+  #define ELPP_DISABLE_LOGS
+  #define ML_NOLOG
+#endif
+
+#include <CORENEURON/easylogging++.h>
+
+//! @{
+#ifndef ML_NOLOG
+  /**
+   * same as `LOG(level, x)` from easylogging++
+   *
+   * @see @ref logging_macros
+   */
+  #define ML_LOG(level, x) LOG(level) << x
+
+  /**
+   * same as `CLOG(level, logger, x)` from easylogging++
+   *
+   * @see @ref logging_macros
+   */
+  #define ML_CLOG(level, logger_id, x) CLOG(level, logger_id) << x
+
+  /**
+   * same as `CLOG_IF(condition, level, logger, x)` from easylogging++
+   *
+   * @see @ref logging_macros
+   */
+  #define ML_CLOG_IF(condition, level, logger_id, x) CLOG_IF(condition, level, logger_id) <<  x
+
+  /**
+   * same as `CVLOG(verbosity, logger, x)` from easylogging++
+   *
+   * @see @ref logging_macros
+   */
+  #define ML_CVLOG(verbose_level, logger_id, x) CVLOG(verbose_level, logger_id) << x
+
+  /**
+   * same as `CVLOG_IF(condition, verbosity, logger, x)` from easylogging++
+   *
+   * @see @ref logging_macros
+   */
+  #define ML_CVLOG_IF(condition, verbose_level, logger_id, x) CVLOG_IF(condition, verbose_level, logger_id) << x
+#else
+  #define ML_LOG(level, x)
```

Comment:
I see here and further below that in your implementation (maybe that's the easylogging way to do it) you heavily rely on macros. I think that's a bad idea. We are trying to modernize this code, moving away from macros and defines as possible. C++ offers other, more powerful and flexible features to build and use APIs.